Issue 1272: Issues

simplypixi opened the pull request:
This fixes #1272

dleszcz commented:
@carmenfan @sanmont3drepo Ready for review

sanmont3drepo commented:
When a issue is created, it reloads the model

![newissuereloads model](https://user-images.githubusercontent.com/42032433/53429843-4934c580-39e5-11e9-91af-38e3f8cbd7fe.gif)

sanmont3drepo commented:
When a issue is being closed it used to show the message that someone was closing it:

Ok:
![closing-issue ok](https://user-images.githubusercontent.com/42032433/53431125-cf520b80-39e7-11e9-856c-7929bfde2671.gif)

wrong:
![closing issue wrong](https://user-images.githubusercontent.com/42032433/53431160-ded15480-39e7-11e9-829e-27dc3d3945ff.gif)

sanmont3drepo commented:
The filter for date is wrong:

ok:
![date from ok](https://user-images.githubusercontent.com/42032433/53431772-299f9c00-39e9-11e9-9e1b-14663f8c7f16.PNG)

wrong:
![date from wrong](https://user-images.githubusercontent.com/42032433/53431805-3623f480-39e9-11e9-8f1b-935779261079.PNG)
![date to wrong](https://user-images.githubusercontent.com/42032433/53431808-38864e80-39e9-11e9-85f9-209679629f9a.PNG)

sanmont3drepo commented:
I can "mock" add a second pin when editing on an already existing issue.

![second pin](https://user-images.githubusercontent.com/42032433/53432119-d417bf00-39e9-11e9-96e4-b8e309a2c829.PNG)

carmenfan commented:
The entire model is hidden on an issue when it shouldn't be:
![1](https://user-images.githubusercontent.com/11945337/53432732-f827d000-39ea-11e9-8261-0c1008bb2c32.gif)

sanmont3drepo commented:
You can add twice or more times the same filter. For adding the Status:Closed more thatn once, use the drop down menu and add it, then add it again by using the autocomplete:

![image](https://user-images.githubusercontent.com/42032433/53433032-8ac86f00-39eb-11e9-8806-f8bdbbb35a0e.png)

sanmont3drepo commented:
Using backspace no longer removes the filter chips:

ok:
![backspace filters ok](https://user-images.githubusercontent.com/42032433/53434152-dd0a8f80-39ed-11e9-9ee2-436f424f8fe6.gif)

wrong:
![backspace filters wrong](https://user-images.githubusercontent.com/42032433/53434163-e267da00-39ed-11e9-8e81-5113252948d4.gif)

sanmont3drepo commented:
When downloading a json after the issues being filtered , it downloads the whole lot instead of the filtered issues:
![image](https://user-images.githubusercontent.com/42032433/53434610-e6482c00-39ee-11e9-8ffc-ab374ccbba67.png)

paloknapo commented:
We need to make the issue details part (flags, description) scroll-able as comments don't show on small screens. Collapse button moves to the bottom of the details edge as shown in this updated sketch:

https://xd.adobe.com/view/259b33f9-11dc-4fac-6fbe-1516fa769c71-53aa/

Also, focus on comment once it's added. It is not visible at the moment if the details are expanded. 

![01](https://user-images.githubusercontent.com/760452/53435750-6ff8f900-39f1-11e9-97d1-6af963fcb35d.gif)

paloknapo commented:
Focus on recently opened issue left when going back to list view:

![01](https://user-images.githubusercontent.com/760452/53436097-48eef700-39f2-11e9-89f6-9719509826f9.gif)

sanmont3drepo commented:
Submodels issues are being shown regardless if the "show submodels issues" filter has being set or not:

Ok:
![federation issues-filters](https://user-images.githubusercontent.com/42032433/53484201-2b19a480-3a7b-11e9-9fe1-b90669d4d45b.gif)

Wrong:
![federation issues-filters wrong](https://user-images.githubusercontent.com/42032433/53484214-31a81c00-3a7b-11e9-8943-c143e9e5524f.gif)

carmenfan commented:
> When downloading a json after the issues being filtered , it downloads the whole lot instead of the filtered issues:
> ![image](https://user-images.githubusercontent.com/42032433/53434610-e6482c00-39ee-11e9-8ffc-ab374ccbba67.png)

Just to clarify on this one, this was an update done whilst you guys already started the work. (Issue here: https://github.com/3drepo/3drepo.io/issues/1254)

The download url should also have the query string to convert coordinates to directX

carmenfan commented:
- [ ] Don't move the camera view when the cross is pressed
![1](https://user-images.githubusercontent.com/11945337/54219008-88cfd700-44e6-11e9-9ee7-94a10ac6d065.gif)

- [ ] We should show date/time on system messages as well
![image](https://user-images.githubusercontent.com/11945337/54219058-a1d88800-44e6-11e9-8244-f4583952a453.png)

- [ ] Show year if the due date is not within the current year 
![1](https://user-images.githubusercontent.com/11945337/54219169-dfd5ac00-44e6-11e9-9353-0162efef1f38.gif)

- [ ] Save button should only enable if there's something to save:
![image](https://user-images.githubusercontent.com/11945337/54219573-b10c0580-44e7-11e9-90f4-977a77c268fd.png)

- [ ] The title strip should not be within the scrollable part. It should always be visible: 
![image](https://user-images.githubusercontent.com/11945337/54220468-64292e80-44e9-11e9-89c3-5aad858cf33a.png)

- [ ] Can we make clicking on the title strip to go to the issue's viewpoint? 
![image](https://user-images.githubusercontent.com/11945337/54220641-c41fd500-44e9-11e9-81d4-c721610ae079.png)

- [ ] Error message if I open an issue -> leave the model -> open another model:
![1](https://user-images.githubusercontent.com/11945337/54221057-9be4a600-44ea-11e9-9a3f-7a40b64da1f3.gif)

simplypixi commented:
> * [ ]  The entire model is hidden on an issue when it shouldn't be:
>   ![1](https://user-images.githubusercontent.com/11945337/53432732-f827d000-39ea-11e9-8261-0c1008bb2c32.gif)

@carmenfan  Is it happening on different models? Are able to provide us more details? I went through each model that we have on the local environment and everything seems to work properly. :(

Charence commented on backend/models/issue.js:
```diff
@@ -525,7 +525,6 @@ schema.statics.getIssuesList = function(dbColOptions, user, branch, revision, id
 
 	const projection = {
 		extras: 0,
-		"comments": 0,
```
Why has this been removed?

simplypixi replied:
Because we need list of comments for each issues to filter them by text included in comment.

Charence replied:
The filter should not check the contents of the comments. It is sufficient to only filter the issue details excluding comments. I have confirmed this with @carmenfan.

simplypixi replied:
That's weird. 😕 So, why do you have this functionality on staging? Do want to remove it?

carmenfan replied:
@simplypixi 

I don't know how react is doing it (@sanmont3drepo is doing the code review), but in the angular version, the issue list shares the data model with issue details

When we fetch the issue list, it doesn't contain comments (as issue list can get overly large, we want to just get the "cliff notes" of each issue, sufficient to show the listing). 

If the user clicked into the issue (i.e. if we've entered the issue detail tab, where full issue details (including comments) are fetched, this updates the data model that is used by both issuelist and issue details

when a user does a string search, we look through that data model. Depending on whether the full issue details has been fetched or not, some issues may have comments loaded and some don't. As of staging, we look through the comments if it's in the datamodel , but if it's not there we don't try to fetch them.

If the data model is separate in the react version (should it though?), I'd remove this feature. I guess the bottom line I want to say is, we don't want to fetch the full issue detail on the issueList fetch API.

simplypixi replied:
I understand. We'll implement this the same way as in the angular version. ;)

simplypixi replied:
Fixed ;]

carmenfan commented:
> @carmenfan Is it happening on different models? Are able to provide us more details? I went through each model that we have on the local environment and everything seems to work properly. :(

@simplypixi I've uploaded a similar (effectively the same but different version) of the model into your database, it's in someCrazilyLongNameLikeThis/8f1fb36b-b95e-4e64-bdd3-4f957e7ceb79 (Again, feel free to change the password)

I also created an issue and the same thing is happening. I have a feeling it may be related to models with IFC Spaces (which are invisible by default, so if you look at the tree you can see there are invisible objects:
![1](https://user-images.githubusercontent.com/11945337/54359331-521acd80-465a-11e9-8827-2c7ee4dfd376.gif)



I actually ran into an error when I saved an issue in this model, not sure if it's related:
![image](https://user-images.githubusercontent.com/11945337/54359360-62cb4380-465a-11e9-9faa-3c7cafb1a1b8.png)

simplypixi commented:
> * [ ]  When a issue is created, it reloads the model
> 
> ![newissuereloads model](https://user-images.githubusercontent.com/42032433/53429843-4934c580-39e5-11e9-91af-38e3f8cbd7fe.gif)

@carmenfan I checked this bug and it's not so easy to fix because of the way how angular loads routes. It does it in a bit a different way than React. Currently, I'm working on a potential solution but I would like to mark that it can take some time before I finish.

simplypixi commented:
@carmenfan Build fails on style's tests. It's related to file, which is only on staging branch. It's happening because of that you probably build staging with merged changes from specific branch, isn't? @dleszcz has prepared other PR with required fixes, so I think we should merge it first into the staging.

Nevertheless, PR is ready. ;)

dleszcz commented:
@carmenfan staging rebased & build fixed

simplypixi commented:
![image](https://user-images.githubusercontent.com/5683490/54738322-129d3580-4bb4-11e9-8b63-b2cd3ab470a1.png)
Just a few conflicts... :D

simplypixi commented:
> ![image](https://user-images.githubusercontent.com/5683490/54738322-129d3580-4bb4-11e9-8b63-b2cd3ab470a1.png)
> Just a few conflicts... :D

@carmenfan @sanmont3drepo @Charence I've rebased this branch (ugh). Like you can see above, there where a lot of conflicts. I think we should test issues and risks one more time to be sure about that everything works properly. I went through them and I found some small bugs (which I also fixed).
But in my opinion, it should be double-checked by someone else.

carmenfan commented:
> @carmenfan @sanmont3drepo @Charence I've rebased this branch (ugh). Like you can see above, there where a lot of conflicts. I think we should test issues and risks one more time to be sure about that everything works properly. I went through them and I found some small bugs (which I also fixed).
> But in my opinion, it should be double-checked by someone else.

Will do!

carmenfan commented:
EDIT: @simplypixi - Struggling to build this, maybe we lost some dependencies?
![image](https://user-images.githubusercontent.com/11945337/54746954-6ebc8580-4bc5-11e9-9874-cae8cee48ebb.png)

simplypixi commented:
> EDIT: @simplypixi - Struggling to build this, maybe we lost some dependencies?
> ![image](https://user-images.githubusercontent.com/11945337/54746954-6ebc8580-4bc5-11e9-9874-cae8cee48ebb.png)

There was a conflict related to this package... I'll check it!

paloknapo commented:
- [ ] Notifications are cleared when Issue flag is changed
![01](https://user-images.githubusercontent.com/760452/54832554-49f30b80-4cb4-11e9-914c-3518162d9e78.gif)

- [ ] Issues card disappears when user clicks on Due Date field

- [ ] Issue out of focus when going back to list view
![01](https://user-images.githubusercontent.com/760452/54832798-c685ea00-4cb4-11e9-95e5-232a9203816b.gif)

- [ ] Scrollbar appears without reason when changing issue flags
![01](https://user-images.githubusercontent.com/760452/54832948-11076680-4cb5-11e9-85e2-fb066ef74c27.gif)

dleszcz commented:
Hi @PavolKnapo 

>  Issues card disappears when user clicks on Due Date field

Did you do any specific steps? Because we can't reproduce this.

> Scrollbar appears without reason when changing issue flags

This one can't be fixed now because of the height panel setting mechanism. It's a temporary solution, and we will be able to fix the scroll issue after rewriting this component. Is it ok for you?

paloknapo commented:
> Hi @PavolKnapo
> 
> > Issues card disappears when user clicks on Due Date field
> 
> Did you do any specific steps? Because we can't reproduce this.

I just opened a model, opened issues card, opened card and clicked on the Due Date. Card disappears and I get this errors:
![image](https://user-images.githubusercontent.com/760452/54923707-b400f000-4f02-11e9-81ba-2fccaad70c7f.png)
![image](https://user-images.githubusercontent.com/760452/54923722-bebb8500-4f02-11e9-8a51-e0e9222a3d3c.png)

![image](https://user-images.githubusercontent.com/760452/54923633-903daa00-4f02-11e9-9c05-59c4d827b1d4.png)


> 
> > Scrollbar appears without reason when changing issue flags
> 
> This one can't be fixed now because of the height panel setting mechanism. It's a temporary solution, and we will be able to fix the scroll issue after rewriting this component. Is it ok for you?

Yes, that's fine.

Thanks.

simplypixi commented:
> I just opened a model, opened issues card, opened issue and clicked on the Due Date. Card disappears and I get this errors:

@paloknapo Fixed ;)

carmenfan commented:
@simplypixi @dleszcz 

We've just merged in Safetibase v2, hence the merge conflicts in this PR. Please don't worry about them,  @Charence will deal with them (and also the [remaining change request](https://github.com/3drepo/3drepo.io/pull/1427#pullrequestreview-218991394) on Safetibase)

Charence commented:
@simplypixi @dleszcz Conflicts resolved!

sanmont3drepo commented on frontend/modules/issues/issues.redux.ts:
```diff
@@ -0,0 +1,182 @@
+
+export const fetchIssueSuccess = (state = INITIAL_STATE, { issue }) => {
+	const issuesMap = cloneDeep(state.issuesMap);
+	issuesMap[issue._id].comments = issue.comments;
```
Deep cloning is being used everywhere in this pr to update state. I think this might not right and will affect perfomance.

from this article:
https://redux.js.org/faq/performance#do-i-have-to-deep-clone-my-state-in-a-reducer-isnt-copying-my-state-going-to-be-slow

![image](https://user-images.githubusercontent.com/42032433/55622043-17a6db00-5797-11e9-9719-d30e5bee2543.png)

How to update state the proper way:
https://redux.js.org/recipes/structuring-reducers/immutable-update-patterns

simplypixi replied:
Good point!  👍

simplypixi replied:
Fixed

simplypixi commented:
@Charence I get this warning always when I add new comment:
![image](https://user-images.githubusercontent.com/5683490/55649315-f79d0900-57e2-11e9-8c54-3d4bbe35f14c.png)

It's related to this part of code:
![image](https://user-images.githubusercontent.com/5683490/55649338-084d7f00-57e3-11e9-9091-b913a34879b7.png)

Charence commented:
> `guid` seems to be undefined

@simplypixi Looking into it!

Charence commented:
@simplypixi I'm trying to resolve the merge conflicts. Which one should I keep for the following?

frontend/helpers/searching.ts:
![Screenshot 2019-04-08 15 05 36](https://user-images.githubusercontent.com/1438332/55730327-d2d9a900-5a0f-11e9-828d-133d2136da63.png)

frontend/routes/components/filterPanel/filterPanel.component.tsx:
![Screenshot 2019-04-08 14 53 12](https://user-images.githubusercontent.com/1438332/55729597-5397a580-5a0e-11e9-84a3-e6906740817a.png)

frontend/routes/viewer/components/gis/components/settingsForm/settingsForm.component.tsx:
![Screenshot 2019-04-08 15 04 36](https://user-images.githubusercontent.com/1438332/55730250-a9b91880-5a0f-11e9-99a0-e2f72e16e047.png)

frontend/routes/viewer/components/gis/gis.component.tsx:
![Screenshot 2019-04-08 15 03 15](https://user-images.githubusercontent.com/1438332/55730164-7e362e00-5a0f-11e9-9479-605175fcac9b.png)

simplypixi commented:
> @simplypixi I'm trying to resolve the merge conflicts. Which one should I keep for the following?
> 
> frontend/helpers/searching.ts:
> ![Screenshot 2019-04-08 15 05 36](https://user-images.githubusercontent.com/1438332/55730327-d2d9a900-5a0f-11e9-828d-133d2136da63.png)
> 
> frontend/routes/components/filterPanel/filterPanel.component.tsx:
> ![Screenshot 2019-04-08 14 53 12](https://user-images.githubusercontent.com/1438332/55729597-5397a580-5a0e-11e9-84a3-e6906740817a.png)
> 
> frontend/routes/viewer/components/gis/components/settingsForm/settingsForm.component.tsx:
> ![Screenshot 2019-04-08 15 04 36](https://user-images.githubusercontent.com/1438332/55730250-a9b91880-5a0f-11e9-99a0-e2f72e16e047.png)
> 
> frontend/routes/viewer/components/gis/gis.component.tsx:
> ![Screenshot 2019-04-08 15 03 15](https://user-images.githubusercontent.com/1438332/55730164-7e362e00-5a0f-11e9-9479-605175fcac9b.png)
> 
> frontend/routes/viewer/components/viewerPanel/viewerPanel.component.tsx:
> ![Screenshot 2019-04-08 15 06 59](https://user-images.githubusercontent.com/1438332/55730402-fef52a00-5a0f-11e9-9e67-387ba896a6e5.png)

I will resolve these conflicts tomorrow morning. ;)

Charence commented:
> > `guid` seems to be undefined
> 
> @simplypixi Looking into it!

@simplypixi Fixed!

simplypixi commented:
@Charence @carmenfan @sanmont3drepo 
I've resolved conflicts and fixed some bugs, which were (probably) a result of merging. I think we should test this PR again... I went through the places, which were related to the conflicts (mainly: filters, comments, measure service). Everything seems to be fine, but it will be better if you will also look at it.
Besides, in my opinion, this PR should be merged as soon as possible - even with small bugs, which can be fixed directly on staging branch.

Let me know what do you think about that. ;)

Charence commented:
@simplypixi I've found 1 minor bug, but I think we can merge this first and fix from staging as you mentioned.

The bug occurs when you load an Issue or Risk directly using the URL query string.
<img width="1680" alt="Screenshot 2019-04-09 10 44 04" src="https://user-images.githubusercontent.com/1438332/55790882-3e298680-5ab5-11e9-960d-24affdb4d98c.png">
